ocaml/xapi/xapi_host.ml

```diff
@@ -2923,6 +2923,76 @@ let emergency_reenable_tls_verification ~__context =
   Helpers.touch_file Constants.verify_certificates_path ;
   Db.Host.set_tls_verification_enabled ~__context ~self ~value:true
 
+(** Issue an alert if /proc/sys/kernel/tainted indicates particular kernel
+    errors. Will send only one alert per reboot *)
+let alert_if_kernel_broken =
+  (* This is an optimization. We only know for certain if the alert has been
+     sent if the toolstack hasn't been restarted. If it was restarted, we need
+     to check all the alerts since last reboot (and issue an alert for the
+     first time if needed *)
+  let which_alerts_have_been_sent =
+    ref
+      [
+        ((4, Api_messages.kernel_is_broken "MCE"), false)
+      ; (* processor reported a Machine Check Exception (MCE) *)
+        ((5, Api_messages.kernel_is_broken "BAD_PAGE"), false)
+      ; (* bad page referenced or some unexpected page flags *)
+        ((7, Api_messages.kernel_is_broken "BUG"), false)
+      ; (* kernel died recently, i.e. there was an OOPS or BUG *)
+        ((9, Api_messages.kernel_is_broken_warning "WARN"), false)
+      ; (* kernel issued warning *)
+        ((14, Api_messages.kernel_is_broken_warning "SOFT_LOCKUP"), false)
+        (* soft lockup occurred *)
+      ]
+  in
+  fun ~__context ->
+    (* Only add an alert if (a) we found a problem and
+       (b) an alert wasn't already issued for the currently booted kernel *)
+    which_alerts_have_been_sent :=
+      List.map
```

Comment:
Wouldn't it be nicer if this was using a filter_map, and dropping the sent items? This also allows to remove the has_alert_been_sent.